fiftyone/utils/beam.py

```diff
@@ -338,6 +345,194 @@ def beam_export(
             )
 
 
+def beam_map(
+    sample_collection,
+    map_fcn,
+    reduce_fcn=None,
+    save=None,
+    shard_size=None,
+    shard_method="id",
+    num_workers=None,
+    options=None,
+    progress=False,
+    verbose=False,
+):
+    """Applies the given function to each sample in the collection via
+     `Apache Beam <https://beam.apache.org>`_.
+
+    When only a ``map_fcn`` is provided, this function is a parallelized
+    version of
+    :meth:`iter_samples(autosave=True) <fiftyone.core.collections.SampleCollection.iter_samples>`
+    that effectively performs the following operation with the outer loop
+    in parallel::
+
+        for batch_view in fou.iter_batches(sample_collection, shard_size):
+            for sample in batch_view.iter_samples(autosave=True):
+                map_fcn(sample)
+
+    When a ``reduce_fcn`` is provided, this function effectively performs the
+    following operation with the outer loop in parallel::
+
+        values = {}
+        for batch_view in fou.iter_batches(sample_collection, shard_size):
+            for sample in batch_view.iter_samples(autosave=save):
+                values[sample.id] = map_fcn(sample)
+
+        output = reduce_fcn(sample_collection, values)
+
+    Example::
+
+        import fiftyone as fo
+        import fiftyone.utils.beam as foub
+        import fiftyone.zoo as foz
+
+        dataset = foz.load_zoo_dataset("cifar10", split="train")
+        view = dataset.select_fields("ground_truth")
+
+        #
+        # Example 1: map
+        #
+
+        def map_fcn(sample):
+            sample.ground_truth.label = sample.ground_truth.label.upper()
```

Comment:
Concerned that by importing fiftyone in the main process we are essentially passing the mongo client to the worker processes instead of independently creating the client inside the worker, which may result in deadlock in real workloads

Reply:
Database connections aren't serializable so it can't be happening.

Reply:
Actually you're correct that mongo clients can be passed to worker processes. However I still don't believe that is happening with `beam_map()` because I **DID** initially see warnings about sharing mongo clients in https://github.com/voxel51/fiftyone/pull/5480 which I had to manually address, but I don't see those errors here.

So it appears that Beam's direct multiprocessing runner has a more localized way of spawning/forking new processes? 🤷

Reply:
According to the documentation, each worker is in a sub process and communicates through gRPC, so they are probably initializing new connection 

https://beam.apache.org/documentation/runners/direct/#:~:text=multi_threading%3A%20Runner%20and%20workers%20communicate,worker%20runs%20in%20a%20thread